### ISSUE TITLE
PEAR.Functions.ValidDefaultValue does not work if the variable is typehinted

### ISSUE THREAD

**Issue by gmponos** [squizlabs/PHP_CodeSniffer#1331]
Using the 2.8.0 version

if I have the following method

```
<?php

public function test(array $var = null, $var2)
{
}
```
Rule `PEAR.Functions.ValidDefaultValue` should catch the case above and display an error because the variable with the default value is before a variable without a default value. But it seems not to be working

**Comment by gmponos** [comment]
Happens also in 2.7.1,

It seems that the problem is a combination of the `null` as default value and the typehint.
If I have this it works fine:

```
<?php

public function test(array $var = [], $var2)
{
}
```

Also this does not work.

```
<?php

public function test(DateTimeInterface $var = null, $var2)
{
}
```

but the codesniffer finds this:

```
<?php

public function test(DateTimeInterface $var = 2, $var2)
{
}
```

**Comment by aik099** [comment]
When you expect `PEAR.Functions.ValidDefaultValue` error to happen?

**Comment by aik099** [comment]
Actually there was a change to that sniff recently and this is expected behavior see: https://github.com/squizlabs/PHP_CodeSniffer/blob/master/CodeSniffer/Standards/PEAR/Sniffs/Functions/ValidDefaultValueSniff.php#L78-L80

That's why `ClassName $paramName = null` isn't considered as optional parameter. I guess this is pre PHP 7.1 way to specify `ClassName ?$paramName` (nullable type) stuff.

**Comment by gmponos** [comment]
I have updated my comments. Is this more clear?

**Comment by aik099** [comment]
The described change was introduced in #1092 ticket.

**Comment by gmponos** [comment]
thank you. I will close it and continue there.

**Comment by gmponos** [comment]
I can not reopen the linked issue.

Based on the discussion we had there I believe it should be reopened. If you think otherwise let me know.

**Comment by gsherwood** [comment]
I can't just change this because using `= null` is the only way before 7.1 to allow NULL to be passed for a type hinted argument.

Unfortunately, it also means that the argument is now optional, unlike when using PHP 7.1 nullable types. The leading `?` does not make the argument optional - it just allows NULL to be passed if desired. This is actually what people are trying to replicate in PHP < 7.1 and why PHPCS was changed to support this coding style.

So if I change this back to how it was, I'll have people complain. If I leave it as it is right now, I'll have people complain. There really isn't any way to satisfy everyone here without making one group do something to change the behaviour of PHPCS.

So the best thing to do is look at the intention of the sniff, which was to ensure that optional arguments go at the end of the argument list so that arguments with default values actually mean something. The classic example of a wasted default value is:
```php
public function test($foo = 'foo', $bar) { ... }
```

The problem with the type hinted NULL arguments is that the default value isn't "wasted" - it means something important. It provides functionality that is otherwise not available to a developer. PHPCS isn't trying to stop a developer deciding what order they want arguments in - it's just trying to make sure default values are useful.

This code is not how I would personally do things:
```php
public function test(array $foo = null, $bar) { ... }
```

I'd write it like this:
```php
public function test($bar, array $foo = null) { ... }
```

And in PHP 7.1, I'd obviously write it like this:
```php
public function test(?array $foo, $bar) { ... }
```

But enforcing this, as the sniff used to do, means that developers have to order their arguments to satisfy PHPCS even though they are using syntax to satisfy PHP *and* PHP doesn't care about default argument values.

So I'm going to leave this issue open, but I'm not going to make any immediate change. I believe that the current functionality of the sniff reflects its intentions, even though it's pretty borderline.

**Comment by gsherwood** [comment]
I'm now closing the issue as there has been no further feedback, and there is no intention to change the behaviour at this time, especially in the 2.x version.
